New events and events corrections

vcoindet opened the pull request:
**Adding the following events :**
- animationstarted
- animationstopped
- zoomchanged

**Improvements on the following events :**
- centerchanged (called through user interaction)
- orienationchanged (called through user interaction, access to oldOrientation)
- layerchanged (access to old and state of the layer)
- layerchanged:visible (access to oldvisibility)
- layerchanged:opacity (access to oldopacity)
- layerchanged:index (access to oldindex)
- layeradded (access to the id of the layer added)

vcoindet commented:
Rebased and Ready

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -18,17 +18,101 @@ import { STRATEGY_MIN_NETWORK_TRAFFIC } from '../../../../Scene/LayerUpdateStrat
+var _handlerZoom;
+
+/* eslint-disable no-console */
+const logStep = function logStep(step) {
```
remove

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -18,17 +18,101 @@ import { STRATEGY_MIN_NETWORK_TRAFFIC } from '../../../../Scene/LayerUpdateStrat
+    },
+});
+var eventCenter = new CustomEvent('centerchanged', {
+    detail: {
```
replace by `detail: {old: {},new: {}},`

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -156,11 +241,20 @@ ApiGlobe.prototype.moveLayerDown = function moveLayerDown(layerId) {
 ApiGlobe.prototype.moveLayerToIndex = function moveLayerToIndex(layerId, newIndex) {
+    eventLayerChanged.detail.visibility.oldVisibility = this.scene.getMap().layersConfiguration.isColorLayerVisible(layerId);
+    eventLayerChanged.detail.visibility.newVisibility = this.scene.getMap().layersConfiguration.isColorLayerVisible(layerId);
+    eventLayerChanged.detail.opacity.oldOpacity = this.scene.getMap().layersConfiguration.getColorLayerOpacity(layerId);
```
factorize

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -374,10 +469,18 @@ ApiGlobe.prototype.setRealisticLightingOn = function setRealisticLightingOn(valu
  */
 
 ApiGlobe.prototype.setLayerVisibility = function setLayerVisibility(id, visible) {
+    eventLayerChanged.detail.index.oldIndex = this.scene.getMap().layersConfiguration.layersState[id].sequence;
```
factorize

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -389,10 +492,18 @@ ApiGlobe.prototype.setLayerVisibility = function setLayerVisibility(id, visible)
  */
 
 ApiGlobe.prototype.setLayerOpacity = function setLayerOpacity(id, opacity) {
+    eventLayerChanged.detail.index.oldIndex = this.scene.getMap().layersConfiguration.layersState[id].sequence;
```
factorize

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -614,14 +735,17 @@ ApiGlobe.prototype.setSceneLoaded = function setSceneLoaded() {
 ApiGlobe.prototype.setCenter = function setCenter(coordinates, isAnimated) {
     isAnimated = isAnimated || this.isAnimationEnabled();
-    eventCenter.oldCenter = this.getCenter();
+    eventCenter.detail.oldCenter.longitude = this.getCenter()._values[0] * 180 / Math.Pi;
```
Math.Pi???

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -614,14 +735,17 @@ ApiGlobe.prototype.setSceneLoaded = function setSceneLoaded() {
         this.scene.notifyChange(1);
         return this.setSceneLoaded().then(() => {
             this.scene.currentControls().updateCameraTransformation();
+            eventCenter.detail.newCenter.longitude = this.getCenter()._values[0] * 180 / Math.Pi;
```
use Coordinates functions

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -817,6 +944,109 @@ ApiGlobe.prototype.setZoomScale = function setZoomScale(zoomScale, pitch, isAnim
     return this.setRange(range, isAnimated);
 };
 
+ApiGlobe.prototype.testEvents = function testEvents(eventname, param) {
```
remove and keep in you side

vcoindet committed:
feat(event) : Adding events and event improvements

Adding the following events :
- animationstarted
- animationstopped
- zoomchanged

Improvements on the following events :
- centerchanged (called through user interaction)
- orienationchanged (called through user interaction, access to oldOrientation)
- layerchanged (access to old and state of the layer)
- layerchanged:visible (access to oldvisibility)
- layerchanged:opacity (access to oldopacity)
- layerchanged:index (access to oldindex)
- layeradded (access to the id of the layer added)

vcoindet commented:
rebased and ready @gchoqueux

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -11,24 +11,53 @@ import WMTS_Provider from '../../Providers/WMTS_Provider';
-var eventCenter = new CustomEvent('centerchanged');
-var eventOrientation = new CustomEvent('orientationchanged');
+var eventRange = new CustomEvent('rangeChanged', {
+    detail: { new: {}, old: {} },
```
create object `statut`
 
detail : new statut(),

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -366,6 +403,32 @@ ApiGlobe.prototype.setRealisticLightingOn = function setRealisticLightingOn(valu
     this.scene.renderScene3D();
 };
 
+ApiGlobe.prototype.getOldStatusLayer = function getOldStatusLayer(layerId) {
```
Only one function is required `getStatusLayer`, no?

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -366,6 +403,32 @@ ApiGlobe.prototype.setRealisticLightingOn = function setRealisticLightingOn(valu
 
+ApiGlobe.prototype.getOldStatusLayer = function getOldStatusLayer(layerId) {
+    eventLayerChanged.detail.index.old = this.scene.getMap().layersConfiguration.layersState[layerId].sequence;
+    eventLayerChanged.detail.opacity.old = this.scene.getMap().layersConfiguration.getColorLayerOpacity(layerId);
```
`eventLayerChanged.detail.opacity.old` or `eventLayerChangedOpacity.detail.old`
need homogenized details

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -843,6 +967,41 @@ ApiGlobe.prototype.callbackLayerChanged = function callbackLayerChanged() {
 };
 
+ApiGlobe.prototype.addEventListenerCenterChanged = function addEventListenerCenterChanged() {
+    this.getCenterStatus('new');
```
and the old statut?

vcoindet replied:
No need here, it's just a initialization so the old status will be good in the function `callbackCenterChanged`

It's the same for the events `zoomchanged` and `orientationchanged`

gchoqueux commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -651,6 +656,15 @@ function GlobeControls(camera, domElement, engine) {
+        targt.y = globeTarget.position.y;
+
+        var rot = {};
+        rot.x = globeTarget.rotation._x;
```
not correct, i think is more `spherical` member

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -865,6 +1033,22 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
+    this.viewerDiv.removeEventListener('orientationcontrolchanged', _handlerOrientation, false);
+};
+
+ApiGlobe.prototype.selectNodeById = function selectNodeById(id) {
```
Why you add this?

autra commented:
2 important points:

- A lot of boilerplate would have been spared if you didn't try to create events as constants... Also doing so is dangerous: sharing them as it is done here (and in itowns currently) is a risk of data leakage and of collision. Just create them when you fire them. Events are not made to be shareable like that. 

It would be simpler and safer and would avoid having stuff like the `status` function, that really does nothing.

*  even if there might be exceptions. ApiGlobe should *not* fire any events : it's proxying stuff to the underlying services (Scene, GlobeControls) etc and *they* should fire events. The point of events is to notify client of changes that itowns is doing without the client knowing *or* to notify them of any asynchronous action that have been completed.

For example: `moveLayerUp` is not asynchrone AND is called *by* the client. The client code (geoportail, or anything else) already *knows* about the change. It simply doesn't make sense to fire an event there. Other examples of useless events : all the setLayer* methods,

autra commented:
> it's proxying stuff to the underlying services (Scene, GlobeControls) etc and they should fire events

*if* there is an asynchrone process of course.

vcoindet committed:
moving events

gchoqueux commented:
moved in #306 and #298
